File: py-usfm-parser/tests/test_parsing.py
'''To test parsing success/errors for USFM/X committee's test suite'''
import pytest
from lxml import etree

from src.usfm_grammar import USFMParser
from tests import all_usfm_files, initialise_parser, is_valid_usfm,\
    negative_tests, find_all_markers

test_files = all_usfm_files.copy()

@pytest.mark.parametrize('file_path', test_files)
def test_error_less_parsing(file_path):
    '''Tests if input parses with or without errors, as expected'''
    test_parser = initialise_parser(file_path)
    if is_valid_usfm(file_path):
        # positive tests
        assert not test_parser.errors, test_parser.errors
    else:
        # negative tests
        # assert test_parser.errors, "file has errors, but passed\n"+test_parser.to_syntax_tree()
        assert test_parser.errors or "MISSING" in test_parser.to_syntax_tree(),\
            "file has errors, but passed\n"+test_parser.to_syntax_tree()

positive_files = test_files.copy()
for file in negative_tests:
    if file in positive_files:
        positive_files.remove(file)

def get_nodes(node):
    '''Recursive function to traverse the syntax tree'''
    nodes = [node.type]
    if len(node.children)>0:
        for child in node.children:
            nodes += get_nodes(child)
    return nodes

@pytest.mark.parametrize('file_path', positive_files)
@pytest.mark.timeout(30)
def test_all_markers_are_in_output(file_path):
    '''Tests if all markers in USFM are present in output also'''
    test_parser = initialise_parser(file_path)
    assert not test_parser.errors, test_parser.errors

    all_markers_in_input = find_all_markers(file_path, keep_number=False)

    all_nodes_in_st = get_nodes(test_parser.syntax_tree)
    for marker in all_markers_in_input:
        if marker.startswith("z"):
            marker = "zNameSpace"
        elif marker in ['qte','qts', 'ts'] or marker.endswith('-e') or marker.endswith("-s"):
            marker = "milestone"
        elif marker in ['xt']:
            marker = "crossref"
        assert marker in all_nodes_in_st, marker

USFM_WITH_ERROR = '''
\\id GEN
\\c 1
\\p
\\v 1 correct verse one
\\v 2 correct verse two
\\p
\\v3 wrong verse
\\c 3
\\v 1 verse in chapter without paragraph
\\p
\\v 2 a correct verse following one without para
\\c 4
\\s5
\\p
\\v 1 correct verse three after s5
'''

def test_partial_parsing_with_errors():
    '''Test use of ignore_errors flag to obtain some output even when input has errors'''
    test_parser = USFMParser(USFM_WITH_ERROR)
    assert test_parser.errors

    # without ignore_errors flag
    def use_API_negative(test_parser, api_str_expression):
        '''negative tests to ensure exception is raised'''
        threw_error = False
        try:
            eval(api_str_expression)
        except Exception as exe:
            assert "Errors present:" in str(exe), api_str_expression
            assert "Use ignore_errors=True" in str(exe)
            threw_error = True
        assert threw_error

    use_API_negative(test_parser, 'test_parser.to_usj()')
    use_API_negative(test_parser, 'test_parser.to_list()')
    use_API_negative(test_parser, 'test_parser.to_usx()')

    # with ignore_errors=True
    def use_API_positive(test_parser, api_str_expression):
        '''positive tests to ensure correct portions are made available in output'''
        output = eval(api_str_expression)
        if isinstance(output, etree._Element):
            str_output = etree.tostring(output).decode('utf-8')
        else:
            str_output = str(output)
        assert "correct verse one" in str_output, api_str_expression
        assert "correct verse two" in str_output, api_str_expression
        assert "correct verse three after s5" in str_output, api_str_expression

    use_API_positive(test_parser, "test_parser.to_usj(ignore_errors=True)")
    use_API_positive(test_parser, "test_parser.to_list(ignore_errors=True)")
    use_API_positive(test_parser, "test_parser.to_usx(ignore_errors=True)")
    

File: rust-usfm-parser-new/tests/test_parsing.rs
//! Rust port of `test_parsing.py`
//! Tests basic parsing success/failure against the USFM/X committee test suite.
mod common;

use std::fs;
use usfm_grammar::USFMParser;

use common::{
    all_usfm_files, find_all_markers, initialise_parser, is_valid_usfm, negative_tests,
};

// ---------------------------------------------------------------------------
// test_error_less_parsing
// ---------------------------------------------------------------------------

/// Mirrors `test_error_less_parsing`.
/// For each test file: valid files must have no parse errors,
/// invalid files must have errors OR contain "MISSING" in the syntax tree.
#[test]
fn test_error_less_parsing() {
    let files = all_usfm_files();
    assert!(!files.is_empty(), "No test USFM files found under ../tests");

    let mut failures: Vec<String> = Vec::new();

    for path in &files {
        let parser = initialise_parser(path);
        let label  = path.display().to_string();

        if is_valid_usfm(path) {
            // Positive test — no errors expected
            if !parser.errors.is_empty() {
                failures.push(format!(
                    "UNEXPECTED ERROR in {label}: {:?}",
                    parser.errors
                ));
            }
        } else {
            // Negative test — errors OR MISSING required
            let tree_str = match (parser.to_syntax_tree(true)) {
                Ok(node) => node.to_sexp(),
                Err(e) => {
                    println!("Expected error in {label}: {:?}", e);
                    continue; // error is expected, so this is a pass
                }
            };
            if !tree_str.contains("MISSING") {
                failures.push(format!(
                    "Expected error or MISSING in syntax tree for {label}, but got none"
                ));
            }
        }
    }

    assert!(
        failures.is_empty(),
        "test_error_less_parsing failures:\n{}",
        failures.join("\n")
    );
}

// ---------------------------------------------------------------------------
// test_all_markers_are_in_output (syntax tree)
// ---------------------------------------------------------------------------

/// Mirrors `test_all_markers_are_in_output` in test_parsing.py.
/// Checks that every marker found in the raw USFM also appears as a node
/// type in the syntax tree, for all *valid* files.
#[test]
fn test_all_markers_are_in_syntax_tree() {
    let neg: std::collections::HashSet<_> = negative_tests().into_iter().collect();
    let positive_files: Vec<_> = all_usfm_files()
        .into_iter()
        .filter(|p| !neg.contains(p))
        .collect();

    let mut failures: Vec<String> = Vec::new();

    for path in &positive_files {
        let parser = initialise_parser(path);
        if !parser.errors.is_empty() {
            continue; // already covered by test_error_less_parsing
        }

        let markers = find_all_markers(path, false, false);
        let tree_str = match parser.to_syntax_tree(false) {
            Ok(node) => node.to_sexp(), // S-expression is more compact than debug
            Err(e) => {
                failures.push(format!(
                    "Failed to get syntax tree for {}: {:?}",
                    path.display(),
                    e
                ));
                continue;
            }
        };

        for raw_marker in &markers {
            // Apply the same synonym map as Python
            let marker = if raw_marker.starts_with('z') {
                "zNameSpace".to_string()
            } else if matches!(raw_marker.as_str(), "qte" | "qts" | "ts")
                || raw_marker.ends_with("-e")
                || raw_marker.ends_with("-s")
            {
                "milestone".to_string()
            } else if raw_marker == "xt" {
                "crossref".to_string()
            } else {
                raw_marker.clone()
            };

            if !tree_str.contains(&marker) {
                failures.push(format!(
                    "Marker '{}' not found in syntax tree of {}",
                    marker,
                    path.display()
                ));
            }
        }
    }

    assert!(
        failures.is_empty(),
        "test_all_markers_are_in_syntax_tree failures:\n{}",
        failures.join("\n")
    );
}

// ---------------------------------------------------------------------------
// test_partial_parsing_with_errors
// ---------------------------------------------------------------------------

const USFM_WITH_ERROR: &str = r#"\id GEN
\c 1
\p
\v 1 correct verse one
\v 2 correct verse two
\p
\v3 wrong verse
\c 3
\v 1 verse in chapter without paragraph
\p
\v 2 a correct verse following one without para
\c 4
\s5
\p
\v 1 correct verse three after s5
"#;

/// Mirrors `test_partial_parsing_with_errors`.
#[test]
fn test_partial_parsing_with_errors() {
    let parser = USFMParser::new(USFM_WITH_ERROR);
    assert!(
        !parser.errors.is_empty(),
        "Expected parse errors for the deliberately broken USFM"
    );

    // Without ignore_errors — all output methods should return Err
    assert!(
        parser.to_usj(None, None, false, true).is_err(),
        "to_usj should fail without ignore_errors"
    );
    assert!(
        parser.to_list(None, None, false, true).is_err(),
        "to_list should fail without ignore_errors"
    );
    assert!(
        parser.to_usx(false).is_err(),
        "to_usx should fail without ignore_errors"
    );

    // Error message must contain the expected hints
    let err_msg = parser.to_usj(None, None, false, true).unwrap_err().to_string();
    assert!(
        err_msg.contains("Errors present:"),
        "Error should mention 'Errors present:'"
    );
    assert!(
        err_msg.contains("ignore_errors"),
        "Error should hint at ignore_errors flag"
    );

    // With ignore_errors=true — correct verses must appear in output
    let usj = parser
        .to_usj(None, None, true, true)
        .expect("to_usj(ignore_errors=true) must succeed");
    let usj_str = usj.to_string();
    assert!(usj_str.contains("correct verse one"),   "USJ missing verse 1");
    assert!(usj_str.contains("correct verse two"),   "USJ missing verse 2");
    assert!(usj_str.contains("correct verse three"), "USJ missing verse 3");

    let list = parser
        .to_list(None, None, true, true)
        .expect("to_list(ignore_errors=true) must succeed");
    let list_str = format!("{:?}", list);
    assert!(list_str.contains("correct verse one"),   "list missing verse 1");
    assert!(list_str.contains("correct verse three"), "list missing verse 3");

    let usx = parser
        .to_usx(true)
        .expect("to_usx(ignore_errors=true) must succeed");
    let usx_str = format!("{:?}", usx); // elementtree::Element debug
    assert!(usx_str.contains("correct verse one"),   "USX missing verse 1");
    assert!(usx_str.contains("correct verse three"), "USX missing verse 3");
}
